Authentication

rishflab opened the pull request:
#229

klochowicz commented on start_all.sh:
```diff
@@ -4,4 +4,4 @@ export RUST_BACKTRACE=1
 # A simple command to spin up the complete package, ie. both daemons and frontends.
 # A single 'ctrl+c' stops all processes.
-(trap 'kill 0' SIGINT; cargo run --bin maker -- testnet & cargo run --bin taker -- testnet & APP=maker yarn --cwd=./frontend dev  & APP=taker yarn --cwd=./frontend dev)
+(trap 'kill 0' SIGINT; cargo run --bin maker -- testnet & cargo run --bin taker -- --maker-id 10d4ba2ac3f7a22da4009d813ff1bc3f404dfe2cc93a32bedf1512aa9951c95e testnet -- & APP=maker yarn --cwd=./frontend dev  & APP=taker yarn --cwd=./frontend dev)
```
will this magic number (`maker_id`) ever change? Perhaps add a comment about it (either a reassuring one that it's set in stone, or how/when it needs to be updated)

klochowicz commented on daemon/src/noise.rs:
```diff
@@ -6,15 +6,19 @@ use tokio::net::TcpStream;
+pub async fn initiator_handshake(
+    connection: &mut TcpStream,
+    local_priv_key: x25519_dalek::StaticSecret,
+    remote_pub_key: x25519_dalek::PublicKey,
```
nit: do you need to own these? ref would probably be enough.

rishflab replied:
yep, ref is enough

klochowicz commented on daemon/src/noise.rs:
```diff
@@ -23,21 +27,22 @@ pub async fn initiator_handshake(connection: &mut TcpStream) -> Result<Transport
-pub async fn responder_handshake(connection: &mut TcpStream) -> Result<TransportState> {
+pub async fn responder_handshake(
+    connection: &mut TcpStream,
+    local_priv_key: x25519_dalek::StaticSecret,
```
nit: do you need to own this? (you could probably avoid a `clone()`)

klochowicz commented on .github/workflows/ci.yml:
```diff
@@ -76,10 +76,12 @@ jobs:
 
-          target/debug/taker --data-dir=/tmp/taker --generate-seed testnet &
+          cp daemon/util/testnet_seeds/taker_seed /tmp/taker/taker_seed:
+          target/debug/taker --data-dir=/tmp/taker --maker-id 10d4ba2ac3f7a22da4009d813ff1bc3f404dfe2cc93a32bedf1512aa9951c95e testnet &
```
will we need to manually upgrade this key, increasing the maintenance burden?

rishflab replied:
if we change the seeds in the repo, yes

rishflab committed:
Authenticate connection with maker

Use the IK noise handshake pattern to authenticate the maker
given knowledge of the maker's static public key.

https://noiseexplorer.com/patterns/IK/

klochowicz commented on daemon/src/taker.rs:
```diff
@@ -51,6 +55,12 @@ struct Opts {
+fn parse_x25519_pubkey(s: &str) -> Result<x25519_dalek::PublicKey> {
+    let mut bytes = [0u8; 32];
+    hex::decode_to_slice(s, &mut bytes)?;
+    Ok(x25519_dalek::PublicKey::from(bytes))
```
nit: `bytes.into()` ?

klochowicz commented on daemon/src/seed.rs:
```diff
@@ -72,6 +72,16 @@ impl Seed {
+            .expand(b"NOISE_STATIC_SECRET", &mut secret)
+            .expect("okm array is of correct length");
+
+        x25519_dalek::StaticSecret::from(secret)
```
nit: `secret.into()`?

klochowicz reviewed (approved):
looks good!
